File: 0037_sudoku_solver/solve_sudoku.java
public class Solution {
    // Time: O(9**n) where n is the number of empty cells
    // Space: O(m * n)
    public void solveSudoku(char[][] board) {
        if (board == null || board.length == 0)
            return;
        solve(board);
    }
    
    private boolean solve(char[][] board){
        int n = board.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] != '.') {
                    continue;
                }
                return tryNumbers(board, i, j);
            }
        }
        return true;
    }
    
    private boolean tryNumbers(char[][] board, int row, int col) {
        for (char c = '1'; c <= '9'; c++){
            if (isValidPosition(board, row, col, c)) {
                board[row][col] = c;
                if (solve(board)) {
                    return true; 
                }
                board[row][col] = '.';
            }
        }
        return false;
    }
    
    private boolean isValidPosition(char[][] board, int row, int col, char c){
        return isValidRow(board, row, c) && 
               isValidCol(board, col, c) &&
               isValidSquare(board, row, col, c);
    }
    
    private boolean isValidRow(char[][] board, int row, char c) {
        for (int i = 0; i < 9; i++) {
            if (board[row][i] == c) {
                return false;
            }
        }
        return true;
    }
    
    private boolean isValidCol(char[][] board, int col, char c) {
        for (int i = 0; i < 9; i++) {
            if (board[i][col] == c) {
                return false;
            }
        }
        return true;
    }
    
    private boolean isValidSquare(char[][] board, int row, int col, char c) {
        for (int i = 0; i < 9; i++) {
            if (board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == c) {
                return false;
            }
        }
        return true;
    }
}

File: 0037_sudoku_solver/solve_sudoku.py
def solve_sudoku(board):
    """
    Time: O(9**n) where n is the number of empty cells
    Space: O(m * n)
    """
    if not board or len(board) == 0:
        return
    solve(board)
        
def solve(board):
    n = len(board)
    for r in range(n):
        for c in range(n):
            if board[r][c] == '.':
                return try_nums(board, r, c)
    return True

def try_nums(board, r, c):
    nums = "123456789"
    for num in nums:
        if is_valid_pos(board, r, c, num):
            board[r][c] = num
            if solve(board): 
                return True
            board[r][c] = '.'
    return False
    
def is_valid_pos(board, r, c, num):
    def is_valid_col():
        for i in range(9):
            if board[i][c] == num: 
                return False
        return True
    def is_valid_row():  
        for i in range(9):
            if board[r][i] == num:
                return False
        return True
    def is_valid_square():
        for i in range(9):
            if board[(r // 3) * 3 + (i // 3)][(c // 3) * 3 + (i % 3)] == num:
                return False    
        return True
    return is_valid_col() and is_valid_row() and is_valid_square()